magicbyirineu/Managers/API/APIManager.swift

```diff
@@ -0,0 +1,84 @@
+import UIKit
+
+class APIManager {
+    // MARK: - Properties
+
+    // MARK: Private
+
+    private let session = URLSession(configuration: .default)
+
+    // MARK: - Init
+
+    // MARK: - Functions
+
+    // MARK: Private
+
+    private func setStatusBar(loading: Bool) {
+        DispatchQueue.main.async {
+            UIApplication.shared.isNetworkActivityIndicatorVisible = loading
+        }
+    }
+
+    private func endpoint<EndpointType: Endpoint>(for request: EndpointType) -> URL {
+        guard let parameters = try? URLQueryEncoder.encode(request) else { fatalError("Wrong parameters") }
+
+        var endpoint = "https://api.magicthegathering.io/v1/\(request.endpoint)"
+
+        if !parameters.isEmpty {
+            endpoint.append("?\(parameters)")
+        }
+
+        let url = URL(string: endpoint)!
+        return url
+    }
+
+    func fetch<EndpointType>(_ request: EndpointType, completion: @escaping (Result<EndpointType.Response>, [AnyHashable: Any]?) -> Void) where EndpointType: Endpoint {
+        setStatusBar(loading: true)
+        let endpoint = self.endpoint(for: request)
+        let task = session.dataTask(with: URLRequest(url: endpoint)) { data, response, error in
+            if let data = data {
+                do {
+                    let decoder = JSONDecoder()
+                    let dateFormatter = DateFormatter()
+                    dateFormatter.dateFormat = "yyyy-MM-dd"
+                    decoder.dateDecodingStrategy = .formatted(dateFormatter)
+                    let magicResponse = try decoder.decode(EndpointType.Response.self, from: data)
+                    let httpResponse = response as? HTTPURLResponse
+
+                    self.setStatusBar(loading: false)
+                    completion(Result.success(magicResponse), httpResponse?.allHeaderFields)
```

Comment:
```suggestion
DispatchQueue.main.async {
       completion(Result.success(magicResponse), httpResponse?.allHeaderFields)
}                   
```

Como vocês estão fazendo a request usando o URLSession, se não me engano o `completion` tem de ser executado na main thread pra vc evitar que lá na frente vc tenha que mandar atualizações de UI na main thread. Em um presenter por exemplo. Todos eles, o de sucesso e o de failure.

Reply:
Faz sentido mesmo 👍